File: desafio3/fsharp/ordenar_vectores.fs
open System
open System.IO
open System.Text


let POS_VECTOR = 9
let ELEMENTOS_VECTOR = 23
let TAM_PERIODO = 6
let CANT_INSTITUCIONES = 6
let TAM_VECTOR = ELEMENTOS_VECTOR * TAM_PERIODO
let TAM_VECTOR_ENTRADA = TAM_VECTOR * CANT_INSTITUCIONES
let LARGO_LINEA = POS_VECTOR + TAM_VECTOR_ENTRADA
let TAM_SALIDA = POS_VECTOR + 1 + TAM_VECTOR

let BUF_SIZE = 65536

let leer nomarch = seq {
    use fr = new StreamReader(new FileStream(nomarch, FileMode.Open, FileAccess.Read, FileShare.Read, BUF_SIZE))
    let mutable nl = 0
    while not fr.EndOfStream do 
        yield (nl, fr.ReadLine())
        nl <- nl + 1
}

let escribir nomarch (lineas : string seq) =
    use fw = new StreamWriter(nomarch, false, Encoding.ASCII,  BUF_SIZE)
    for linea in lineas
        do fw.WriteLine linea

let no_es_cero  (linea:string) (pos:int) = 
    let mutable result = false
    let mutable i = pos
    let top = pos+TAM_PERIODO-1
    while i < top && result = false do
        if linea.[i] <> '0' then 
            result <- true
        i <- i + 1
    result

let separar_periodos (linea:string) = seq {
    let mutable p = POS_VECTOR
    while p < LARGO_LINEA do
        if no_es_cero linea p then
            yield linea.Substring(p, TAM_PERIODO)
        p <- p + TAM_PERIODO
}
    
let PAD_SIZE = TAM_VECTOR+1

let ordenar_periodos (linea:string) = 
    let periodos = separar_periodos linea |> Seq.distinct |> Seq.toList  

    let len = Seq.length periodos
    if len = 0 then "N".PadRight(PAD_SIZE)
    else if len > ELEMENTOS_VECTOR then "S".PadRight(PAD_SIZE)
    else ("D" + (periodos |> Seq.sortDescending |> String.Concat)).PadRight(PAD_SIZE)


let filtrar_linea (n:int, linea : string)=
    if String.length linea <> LARGO_LINEA then 
        printfn "error en linea: %d" n
        linea
    else 
        linea.Substring(0, POS_VECTOR) + ordenar_periodos (linea)


let procesar_vectores entrada salida =
    try
        leer entrada |> Seq.map filtrar_linea |> escribir salida
    with
        | :? System.IO.FileNotFoundException -> 
            printfn "no pudo abrir archivo %s"  entrada

[<EntryPoint>]
let main(argv: string[]) = 
    if argv.Length <> 2 then
        printfn "uso: ordenar-vector archivo_entrada archivo_salida"
    else
        let stopWatch = System.Diagnostics.Stopwatch.StartNew()
        procesar_vectores argv.[0] argv.[1]
        stopWatch.Stop()
        printfn "tiempo ocupado: %d:%d:%d.%d" stopWatch.Elapsed.Hours stopWatch.Elapsed.Minutes stopWatch.Elapsed.Seconds  stopWatch.Elapsed.Milliseconds

    0

File: desafio3/elixir/lib/ordenar_vectores.ex
defmodule Vectores do

	@largo_linea 838
	@pos_vector 9
	@largo_encabezado 9
	@largo_periodo 6
	@largo_vector 828
	@tam_vector 23
	@relleno "      "
	@relleno_largo String.duplicate(@relleno, @tam_vector)

	def main(argv) do
    	{time, _} = :timer.tc(fn () -> 
    		argv |> parse_args |> process 
    	end)

    	"tiempo ocupado: #{time/1_000_000.0} segundos" |> IO.puts
	end

	def parse_args(args) do
		if length(args) !== 2 do 
			:bad_args
		else 
			{:good_args, args}
		end
	end

	def process(:bad_args) do
		"uso: ordenar_vector archivo_entrada archivo_salida" |> IO.puts
	end

	def process({:good_args, [entrada,salida]}) do
		if not File.exists?(entrada) do
			"Archivo: #{entrada} no existe!" |> IO.puts
		else
			File.stream!(entrada)
				|> Stream.with_index
				|> Stream.map(&filtrar_linea/1)
				|> Stream.into(File.stream!(salida))
				|> Stream.run
		end
	end

	defp filtrar_linea({linea, n}) do
		if byte_size(linea) != @largo_linea do
			"Error largo incorrecto en linea #{n}" |> IO.puts
			linea
		else 
			encabezado = binary_part(linea, 0, @largo_encabezado)
			resto = binary_part(linea, @pos_vector, @largo_vector)
			vector = ordenar_vector(resto, 0, @largo_vector, MapSet.new)
			largo = MapSet.size(vector)
			cond do
				largo == 0 -> [encabezado, "N", @relleno_largo, "\n"]
				largo > @tam_vector -> [encabezado, "S",  @relleno_largo, "\n"]
				true -> [encabezado, "D", 
						vector |> MapSet.to_list |> Enum.sort(&(&2<&1)), 
						String.duplicate(@relleno, @tam_vector-largo), "\n"]
			end
		end
	end

	defp ordenar_vector(linea, pos, largo, periodos) do
		periodo = binary_part(linea, pos, @largo_periodo)
		if largo == @largo_periodo do
			agregar_periodo(periodo, periodos)
		else
			ordenar_vector(linea, pos+@largo_periodo, largo-@largo_periodo, agregar_periodo(periodo,periodos))
		end
	end

	defp agregar_periodo(periodo, periodos) do
		if periodo === <<"000000">> do
			periodos
		else
			MapSet.put(periodos, periodo)
		end
	end
end
